[git-webkit] Pull branch name from commit message

JonWBedard opened the pull request:
#### 00b8477c7f93324fb03a3adb886ad0a4f3d6a72f
<pre>
[git-webkit] Pull branch name from commit message
<a href="https://bugs.webkit.org/show_bug.cgi?id=245675">https://bugs.webkit.org/show_bug.cgi?id=245675</a>
&lt;rdar://problem/100412099&gt;

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
(Commit.issues): Parse the commit message for issues.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.title_for): Handle case where commit message is empty.
(PullRequest.pull_request_branch_point): Extract the associated issue from the commit message of a local
commit, if possible.
(PullRequest.is_revert_commit): Handle case where commit message is empty.
(PullRequest.add_comment_to_reverted_commit_bug_tracker): Use Commit object&apos;s issues list.
(PullRequest.create_pull_request): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/commit_unittest.py:
(TestCommit):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:

Canonical link: <a href="https://commits.webkit.org/255245@main">https://commits.webkit.org/255245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c8b904b7e38ba443b8d66af574e05797879e42e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91890 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1134 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/101555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95891 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/1132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97548 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/1132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/95464 "webkitpy-tests (failure)") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/1132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/35990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/90949 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/33727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/37600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1640 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/39484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
```diff
@@ -278,6 +279,24 @@ def uuid(self):
+        seen_empty = False
+        for line in self.message.splitlines():
+            if not line and seen_empty:
+                break
```
Does this mean we may miss some issues from commit message if the commit message contains two consecutive blank lines? For example:
```
Issue Title
Issue-from-tracker1


issue-from-tracker2
```

JonWBedard replied:
Yes, and that's the idea. What we're actually optimizing for is this:

```
Issue title
<issue-from-tracker1>
<issue-from-tracker2>

Reviewed by Ronald Reviewer.

This fixes a problem introduced when fixing <some other issue> in 1234@main....
```

We don't want to include the `<some other issue>`.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
```diff
@@ -278,6 +279,24 @@ def uuid(self):
+                seen_empty = True
+                continue
+            for word in line.split():
+                candidate = Tracker.from_string(word)
```
Do we expect the line most likely only contain one word (e.g. https://github.com/issues/xxxx)? Will it be expensive while processing the title line?

JonWBedard replied:
`<problem> <title>` is a common pattern, as is something like `Cherry-pick 1234@main (abcdef01). <problem>`

I suppose we could simplify this, though, checking the first and last word in a line would seem sufficient.

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[git-webkit] Pull branch name from commit message
https://bugs.webkit.org/show_bug.cgi?id=245675
<rdar://problem/100412099>

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
(Commit.issues): Parse the commit message for issues.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.title_for): Handle case where commit message is empty.
(PullRequest.pull_request_branch_point): Extract the associated issue from the commit message of a local
commit, if possible.
(PullRequest.is_revert_commit): Handle case where commit message is empty.
(PullRequest.add_comment_to_reverted_commit_bug_tracker): Use Commit object's issues list.
(PullRequest.create_pull_request): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/commit_unittest.py:
(TestCommit):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:

Canonical link: https://commits.webkit.org/255245@main
